### PR TITLE
Support jigs for synchronous testing of `zcf` and contract internals

### LINKED ISSUE #1727: Synchronous test jigs for `zcf` and contract testing
**What is the Problem Being Solved?**

To enable testing of ZCF and synchronous updates within a contract, contracts launched 
by a tester must be able to provide direct access to objects within the contract, including
the `zcf` contract facet. 

**Description of the Design**

ZCF is extended with the `setTestJig` operation, to which it can provide a function that will only be invoked when run in a test environment.  That function returns an object to the test that launched the contract. The returned "jig" can hold properties for all the relevant access that the test needs. 

The test provides a function (called `setJig` here) that will be invoked with the above `testJig`, extended 
with a `zcf` argument to provide synchronous access to the contract's `zcf` contract facet. Typically, the 
`setJig` will simply set a variable in teh test environment. For fancier uses, it could extract or merge the 
properties into an overall test jig, allowing multiple contracts to contribute to the overall test access for 
integration tests, for example.

**Security Considerations**
The contract provides access to internal state for testing, but:
- MUST NOT provide that state in production
- MUST NOT receive state from the tester 

**Test Plan**

Integrated into the ZCF and Soe test helpers code. Also used with an actual async membrane in the stablecoin dapp.

### PR DESCRIPTION
**Pull request by dtribble**
Close: #1727
Refs: #1668

Implement the "test jig" design that uses a function to pass the test jig from the contract to the tests. Add it to the ZCF and zoe helper tests.

### CODE REVIEW

File: packages/zoe/src/contractFacet/contractFacet.js [hunk 1]
```diff
@@ -306,10 +329,10 @@ export function buildRootObject() {
         /** @type {PromiseRecord<ZoeSeatAdmin>} */
         const zoeSeatAdminPromiseKit = makePromiseKit();
         // Don't trigger Node.js's UnhandledPromiseRejectionWarning
-        zoeSeatAdminPromiseKit.promise.catch(_ => {});
+        zoeSeatAdminPromiseKit.promise.catch(_ => { });
         const userSeatPromiseKit = makePromiseKit();
         // Don't trigger Node.js's UnhandledPromiseRejectionWarning
-        userSeatPromiseKit.promise.catch(_ => {});
+        userSeatPromiseKit.promise.catch(_ => { });
         const seatHandle = makeHandle('SeatHandle');
```
Comment (katelynsills) [inline packages/zoe/src/contractFacet/contractFacet.js]: why this change?

File: packages/zoe/test/unitTests/contracts/fakeVatAdmin.js [hunk 2]
```diff
@@ -3,23 +3,35 @@ import { makePromiseKit } from '@agoric/promise-kit';
...
+      throw Error(`createVatByName not supported in fake mode`);
+    },
+  });
+  return admin;
+}
+
+const fakeVatAdmin = makeFakeVatAdmin();
+
+export default fakeVatAdmin;
+export { makeFakeVatAdmin };
```
Comment (katelynsills) [inline packages/zoe/test/unitTests/contracts/fakeVatAdmin.js]: needs newline?

File: packages/zoe/test/unitTests/zcf/test-zcf.js [hunk 3]
```diff
@@ -25,9 +27,7 @@ const setupZCFTest = async (issuerKeywordRecord, terms) => {
     terms,
   );
   // This contract gives ZCF as the contractFacet for testing purposes
```
Comment (katelynsills) [inline packages/zoe/test/unitTests/zcf/test-zcf.js]: Outdated comment

File: packages/zoe/test/unitTests/zcf/test-zcfSeat.js [hunk 4]
```diff
@@ -33,9 +34,10 @@ test(`zoe - zcfSeat.kickOut() doesn't throw`, async t => {
     issuerKeywordRecord,
   );
 
-  // This contract gives ZCF as the contractFacet for testing purposes
+  // The contract uses the tstJig so the contractFacet
```
Comment (katelynsills) [inline packages/zoe/test/unitTests/zcf/test-zcfSeat.js]:
```suggestion
  // The contract uses the testJig so the contractFacet 
```

**Review by katelynsills — APPROVED** [review]
Some unrelated changes, but looks good to me!

File: packages/zoe/src/contractFacet/contractFacet.js (hunk 1 above)
Comment (dtribble) [reply packages/zoe/src/contractFacet/contractFacet.js]: That was from running the wrong lint-reformatter. I'll re-run `lint-fix`

File: packages/zoe/src/contractFacet/contractFacet.js [hunk 5]
```diff
@@ -350,6 +350,28 @@ export function buildRootObject() {
...
+       * will be appended to the returned jig object (overriding any
+       * provided by the `testFn`).
+       *
+       * @type SetTestJig
+       */
+      setTestJig: (testFn = () => ({})) => {
+        if (testJigSetter) {
+          console.warn('TEST ONLY: capturing test data', testFn);
+          testJigSetter({ ...testFn(), zcf });
+        }
```
Comment (warner) [inline packages/zoe/src/contractFacet/contractFacet.js:373]: I think this is a bit more complex than necessary, but maybe there are just use cases that aren't exercised by the tests added in this particular PR.

I'm thinking that `setTestJig()` should accept an object as its argument, and call `testJigSetter({...obj, zcf})` rather than taking a function. When I see functions passed into a method like this, I assume that either the caller wants to get something back (but `testFn()` takes no arguments), or they need to know *when* something has happened (but here, it always happens right away), or they need to know *whether* something has happened or not (exposed here, but maybe shouldn't be), or they have a lot of work to do and want to avoid it unless someone really cares (which I don't think is the case).

It might be nice if `setTestJig` were only exposed on the return value if `testJigSetter` was provided, but maybe your intention is to prevent contracts from being able to tell whether it was provided or not (by always providing it, but calling it has no effect unless `testJigSetter` was there).

It might be nice to call `setTestJig(zcf)` immediately, rather than relying upon some test contract to call it later. But obviously you plan to let contracts provide other data too. What about having the third `buildRootObject` be `testJig`, have this section call `testJig.setZCF(zcf)` right away (maybe `testJig?.setZCF(zcf)` is the modern/future way to guard against `testJig===undefined`?), and then wrap and share a `testJig.setContractData()` into the exported object for contracts to (optionally) share additional details if/when they care to?

File: packages/zoe/test/unitTests/zcf/zcfTesterContract.js [hunk 6]
```diff
@@ -7,7 +7,9 @@ import '../../../exported';
  * @type {ContractStartFn}
  */
 const start = zcf => {
-  return { creatorFacet: zcf };
+  // make the `zcf` available to the tests
+  zcf.setTestJig();
```
Comment (warner) [inline packages/zoe/test/unitTests/zcf/zcfTesterContract.js:11]: this call could be removed if you did an automatic `testJig.setZCF(zcf)` as described above

File: packages/zoe/test/unitTests/contracts/fakeVatAdmin.js [hunk 7]
```diff
@@ -3,23 +3,40 @@ import { makePromiseKit } from '@agoric/promise-kit';
...
-          kit.promise.catch(_ => {});
-          return kit.promise;
+function makeFakeVatAdmin(testContextSetter = undefined, makeRemote = x => x) {
+  // This is explicitly intended to be mutable so that
+  // test-only state can be provided from contracts
+  // to their tests.
+  const admin = harden({
+    createVat: bundle => {
+      return harden({
+        root: makeRemote(
```
Comment (warner) [inline packages/zoe/test/unitTests/contracts/fakeVatAdmin.js:13]: this `makeRemote` doesn't appear to be used, is there some future patch that needs it, or was it just for debugging?

**Review by warner — APPROVED** [review]
General approach is good, but I think it could be simplified as described. Feel free to land anyway if you disagree (and once you fix the lint issue).

File: packages/zoe/test/unitTests/contracts/fakeVatAdmin.js (hunk 7 above)
Comment (dtribble) [reply packages/zoe/test/unitTests/contracts/fakeVatAdmin.js:13]: It is used in the dapp-token-economy.  Once this is landed, I'll ty it in these tests (but expect to encounter issues).

File: packages/zoe/src/contractFacet/contractFacet.js (hunk 5 above)
Comment (dtribble) [reply packages/zoe/src/contractFacet/contractFacet.js:373]: The reason for taking a function rather than an argument is so that production code doesn't do any work of setting up for testing if it's in production mode. While that will often just be passing in some variables, it might involve more effort that would be inappropriate in production context. With the current approach, you could imagine the babel transform stripping `setTestJig` invocations.

The reason we don't want to play games with whether `setTestJig` is exposed is that would just require there to be more test-only code in the production contract to see whether it's supposed to be setting up test info.

By having the invocation explicit and not in the return value, more things could be added to the jig over time (e.g., each offer might add something), it can be stripped automatically without changing the visible interface of the contract, and it doesn't go into the replay logs or other persistent state in production. Test jigs are internal details of the module.

File: packages/zoe/test/unitTests/zcf/zcfTesterContract.js (hunk 6 above)
Comment (dtribble) [reply packages/zoe/test/unitTests/zcf/zcfTesterContract.js:11]: Aye. It's fine for it to be explicit, but I can imagine making it always available. That's easy to change in an upwards compatible fashion, so worth discussion even after landing.  Mostly it will depend on how much we use the jig feature.
